src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs

```diff
@@ -130,5 +133,51 @@ protected override Task WriteResponseHeadersAsyncCore(Metadata responseHeaders)
 
             return HttpContext.Response.Body.FlushAsync();
         }
+
+        private DateTime GetDeadline()
+        {
+            const long TicksPerMicrosecond = 10;
+            const long NanosecondsPerTick = 100;
+
+            if (!HttpContext.Request.Headers.TryGetValue(GrpcProtocolConstants.TimeoutHeader, out var values))
+            {
+                return DateTime.MaxValue;
+            }
+
+            if (values.Count == 1)
```

Comment:
nit: maybe the logic for parsing the timeout header shouldn't be directly in the ServerCallContext?